docs/proposals/20190709-cluster-spec-crds.md

```diff
@@ -0,0 +1,325 @@
+    |                 |                             |  | IF Infrastructure has   | |             |
+    |                 |  Get Cluster                |  |    owner ref            | |             |
+    |                 |<----------------------------+--+-------------------------| |             |
+    |                 |                             |  |                         | |--+          |
+    |                 |                             |  |                         | |  | Provision|
+    |                 |                             |  |                         | |  | infra-   |
+    |                 | Set InfrastructureStatusRef |  |                         | |<-+ structure|
+    |                 |<----------------------------+--+-------------------------| |             |
+    |                 |                             |  |                         | |             |
+    |                 |                             |  +-----------------------------------------+
+    |                 |                             |                            +-+
+    |                 |  Cluster update             |                             |
+    |                 |---------------------------> |                             |
+    |                 |                            +-+                            |
+    |                 |  +-----------------------------------------+              |    
+    |                 |  | IF InfrastructureStatusRef              |              |    
+    |                 |  |    is set               | |             |              |    
+    |                 |  |                         | |             |              |    
+    |                 |  | Set InfrastructureStage | |             |              |    
+    |                 |  | to "Provisioned"        | |             |              |    
+    |                 |<-+-------------------------| |             |              |    
+    |                 |  |                         | |             |              |    
+    |                 |  +-----------------------------------------+              |
+    |                 |                            +-+                            |
+    |                 |                             |                             |
+```
+Figure 1: Cluster provisioning process
+
+Figure 1 presents the sequence of actions involved in provisioning a cluster, highlighting the coordination required between the CAPI cluster controller and the provider infrastructure controller. 
+
+The creation of the Cluster and provider infrastructure objects are independent events. When a Cluster object is created, if the `InfrastructureRef` is not set, the cluster controller will set the `InfrastructureState` to "Pending". When a provider infrastructure object is created, the provider's controller will do nothing unless its owner reference is set to a cluster. 
```

Comment:
For v1alpha2 `Machine` status, we do not track the various states of the machine's provider-specific infrastructure. While we do track the machine as pending, provisioning, etc., from an infrastructure perspective, the only status field we have for it is `InfrastructureReady`, which is a boolean.

Reply:
The "pending" status is not related to internal provider state transitions.  It basically states: "this cluster requires an infrastructure but the reference has not yet been set". Actually, the provider cannot start provisioning at this point, as there's no owner reference from the infrastructure object to the cluster. 

If, as suggested in other comments, it is assumed that the Cluster object will have the reference set at creation time, the pending status is not necessary.

Reply:
If you feel like the state is potentially useful, I won't hold this up over that. But I don't think it's necessary. A cluster infra provider should only act (reconcile network, etc.) if the provider resource has an ownerRef back to a Cluster. I believe that is the only prerequisite, and once the ownerRef exists, the infra provider is good to go.

Reply:
Also, it's fine to create the Cluster resource and the infra resource in any order. Once the Cluster's InfrastructureRef is set, then things get set in motion (the cluster controller sets the ownerRef in the infra resource).

Reply:
(unless we decide that InfrastructureRef is required at Cluster creation time)

Reply:
> If you feel like the state is potentially useful,
I don't have a strong opinion on this. I'm just more in the line of making states explicit than requiring external entities (e.g. clients) to reason about the the state by analyzing the Cluster object and/or infrastructure object. I would prefer the controller to encapsulate this logic and be sure every client will get the same result.

Reply:
I'd say cluster.status.infrastructureReady true/false is a good start. The only thing I can see missing there that someone might want to know is "is my infrastructure being reconciled/provisioned?" Is that something that we want to record on the Cluster resource itself?

Reply:
It's probably a good idea to record that on the Cluster itself, since there is no other shared common object that could hold the state.

Reply:
How will we know that it's being provisioned? The presence of (Cluster, Infra resource) doesn't necessarily mean the infra controller is doing anything (yet).